watch-library/hardware/watch/watch_rtc.c

```diff
@@ -22,53 +22,108 @@
     _sync_rtc();
 
-    RTC->MODE2.CTRLA.bit.SWRST = 1;
+    RTC->MODE0.CTRLA.bit.SWRST = 1;
     _sync_rtc();
 
-    RTC->MODE2.CTRLA.bit.MODE = RTC_MODE2_CTRLA_MODE_CLOCK_Val;
-    RTC->MODE2.CTRLA.bit.PRESCALER = RTC_MODE2_CTRLA_PRESCALER_DIV1024_Val;
-    RTC->MODE2.CTRLA.bit.CLOCKSYNC = 1;
-    RTC->MODE2.CTRLA.bit.ENABLE = 1;
+    RTC->MODE0.CTRLA.bit.MODE = RTC_MODE0_CTRLA_MODE_COUNT32_Val;
+    RTC->MODE0.CTRLA.bit.PRESCALER = PRESCALER_SETTING;
+    RTC->MODE0.CTRLA.bit.COUNTSYNC = 1;
+    RTC->MODE0.CTRLA.bit.ENABLE = 1;
+    RTC->MODE0.INTENSET.reg = RTC_MODE0_INTENSET_OVF;
     _sync_rtc();
 }
 
+static const int TZ_OFFSET_DIVIDER = 15;  // i.e. the granularity of our TZ_OFFSET is in 15 minute intervals
+// TODO Mark this as used in movement.h description
+static const int TB_BKUP_REG = 7;
+
+union time_backup {
+    struct {
+        int32_t tz_offset : 8;
+        uint32_t offset : 24;
+    } bit;
+
+    uint32_t reg;
+};
+
 void watch_rtc_set_date_time(watch_date_time date_time) {
-    _sync_rtc(); // Double sync as without it at high Hz faces setting time is unrealiable (specifically, set_time_hackwatch)
-    RTC->MODE2.CLOCK.reg = date_time.reg;
+    union time_backup tb;
+    tb.reg = watch_get_backup_data(TB_BKUP_REG);
+    watch_rtc_set_unix_time(watch_utility_date_time_to_unix_time(date_time, tb.bit.tz_offset * TZ_OFFSET_DIVIDER));
+}
+
+void watch_rtc_set_unix_time_raw(uint64_t unix_time_raw) {
+    union time_backup tb;
+    tb.reg = watch_get_backup_data(TB_BKUP_REG);
+    tb.bit.offset = (unix_time_raw >> 8) / CLK_RTC_CNT_HZ;
+    watch_store_backup_data(tb.reg, TB_BKUP_REG);
     _sync_rtc();
+    RTC->MODE0.COUNT.reg = unix_time_raw & ((1 << 8) * CLK_RTC_CNT_HZ - 1);
 }
 
-watch_date_time watch_rtc_get_date_time(void) {
-    watch_date_time retval;
+void watch_rtc_set_unix_time(uint32_t unix_time) {
+    watch_rtc_set_unix_time_raw((uint64_t)unix_time * CLK_RTC_CNT_HZ);
+}
+
+void watch_rtc_set_tz_offset(uint32_t tz_offset) {
+    union time_backup tb;
+    tb.reg = watch_get_backup_data(TB_BKUP_REG);
+    tb.bit.tz_offset = tz_offset / TZ_OFFSET_DIVIDER;
+    watch_store_backup_data(tb.reg, TB_BKUP_REG);
+}
+
+uint32_t watch_rtc_get_unix_time(void) {
+    union time_backup tb;
+    tb.reg = watch_get_backup_data(TB_BKUP_REG);
+    _sync_rtc();
+    return (tb.bit.offset << 8) + RTC->MODE0.COUNT.reg / CLK_RTC_CNT_HZ;
+}
 
+uint64_t watch_rtc_get_unix_time_raw(void) {
+    union time_backup tb;
+    tb.reg = watch_get_backup_data(TB_BKUP_REG);
     _sync_rtc();
-    retval.reg = RTC->MODE2.CLOCK.reg;
+    return (uint64_t)(tb.bit.offset << 8) * CLK_RTC_CNT_HZ + RTC->MODE0.COUNT.reg;
+}
 
-    return retval;
+watch_date_time watch_rtc_get_date_time(void) {
+    union time_backup tb;
+    tb.reg = watch_get_backup_data(TB_BKUP_REG);
+    _sync_rtc();
+    return watch_utility_date_time_from_unix_time((tb.bit.offset << 8) + RTC->MODE0.COUNT.reg / CLK_RTC_CNT_HZ, tb.bit.tz_offset * TZ_OFFSET_DIVIDER);
```

Comment:
I am concerned about the effect this is going to have on battery life. Doing the unix time -> datetime conversion is a lot of math, and it's going to be called frequently by a lot of faces that need access to the current time. Whenever possible we should err on the side of dedicated hardware doing the math.

Reply:
I understand your concern and I agree. I think I have a solution though: let's make the current time independent of the `RTC`. We can make it part of movement's state and have a high priority 1 Hz task keep the current time every second.

That way we can increase movement's flexibility. We should be able to make the epoch configurable too. I speculate it would also reduce power usage: when the watch doesn't have any background tasks scheduled, this will be the only scheduled task running, and provided we can make the timer facility efficient, it should be much faster than all the loops in the current implementation. And who knows, maybe one day someone will want to compile a firmware _without_ the clock face. Movement can allow any face to act as the low energy watch face.

---

As an aside, I'm also worried about how widely used the `%` modulo operator is in the sensor watch code base. It's famous for being an inefficient operation and it probably uses a lot of power. For example, hash tables typically require the number of buckets to be a power of 2 in order to make modular arithmetic over them efficient.

Reply:
Some thoughts:
- there are a lot of faces call watch_utility_date_time_from_unix_time, which this approach makes much less math-y. It might be interesting to do some kind of survey of what current watch faces need (e.g. some need intervals, some need local time, some just need UTC time (i.e. unix time)). For me the attraction of this approach is that we have a single path to optimise where we might be able to provide _all_ of these use-cases with a high-precision time rather than making trade-offs.
- might be worth experimenting with some kind of caching to optimise the basic clock face time stuff (incrementing only seems scary)

Reply:
I like the caching approach. It would be easy to clear on a 1hz tick interval.

Reply:
I still think most of the time, the watch is going to be displaying a regular clock face, in which case it would be best to take advantage of the RTC's calendar functions, and to synchronize the RTC with the local time zone, not UTC. That would let us save the most power by taking advantage of the hardware and avoiding a bunch of timestamp/timezone math.

The best solution would probably be to leave it up to the user. We could move this change behind a preprocessor flag in `movement_config.h` that lets them switch between "RTC in local time" and "RTC in unix timestamp". For people who want a regular clock face on their watch most of the time, they'd leave the RTC as is, but if their default face is something that makes extensive use of different time zones or unix time, they could switch it to this mode instead.